hamilton/plugins/jupyter_magic.py

```diff
@@ -163,95 +194,160 @@ def determine_notebook_type() -> str:
+          return (A % 3) > 2
+        ```
         """
-        if "--help" in line.split():
-            print("Help for %%cell_to_module magic:")
-            print("  -m, --module_name: Module name to provide. Default is jupyter_module.")
-            print("  -c, --config: JSON config string, or variable name containing config to use.")
-            print("  -r, --rebuild-drivers: Flag to rebuild drivers.")
-            print("  -d, --display: Flag to visualize dataflow.")
-            print("  -v, --verbosity: of standard output. 0 to hide. 1 is normal, default.")
-            return  # Exit early
-        if not hasattr(self, "notebook_env"):
-            # doing this so I don't have to deal with the constructor
-            self.notebook_env = determine_notebook_type()
         # shell.ex() is equivalent to exec(), but in the user namespace (i.e. notebook context).
         # This allows imports and functions defined in the magic cell %%cell_to_module to be
         # directly accessed from the notebook
         self.shell.ex(cell)
 
         args = parse_argstring(self.cell_to_module, line)  # specify how to parse by passing method
-        module_name = "jupyter_module" if args.module_name is None else args.module_name
-
-        display_config = {}
-        if args.config:
-            if args.config in self.shell.user_ns:
-                display_config = self.shell.user_ns[args.config]
+        module_name = args.module_name
+        builder = self.shell.user_ns[args.builder] if args.builder else driver.Builder()
+        inputs = self.shell.user_ns[args.inputs] if args.inputs else {}
+        overrides = self.shell.user_ns[args.overrides] if args.overrides else {}
+        display_config = (
+            self.shell.user_ns[args.display] if args.display not in [True, None] else {}
+        )
+
+        if args.write_to_file:
+            if isinstance(args.write_to_file, str):
+                file_path = Path(args.write_to_file)
             else:
-                try:
-                    if args.config.startswith("'") or args.config.startswith('"'):
-                        # strip quotes if present
-                        args.config = args.config[1:-1]
-                    display_config = json.loads(args.config)
-                except json.JSONDecodeError:
-                    print("Failed to parse config as JSON. Please ensure it's a valid JSON string:")
-                    print(args.config)
-
-        module_object = create_module(cell, module_name)
-
-        # shell.push() assign a variable in the notebook. The dictionary keys are variable name
+                file_path = Path(f"{module_name}.py")
+            file_path.write_text(cell)
+
+        # create modules and build driver
+        module_object = ad_hoc_utils.module_from_source(cell, module_name)
         self.shell.push({module_name: module_object})
+        base_dr = builder.build()  # easier to rebuild a Driver than messing with Builder
```

Comment:
🤔  what's the difficulty? You do need to copy it -- which is something that could be easily added.

```python
base = builder.copy()
base.with_modules(module_object) # this will add to whatever modules were there
# reload any other modules here.
dr = base.build()
```

Reply:
As far as I can tell `Builder.copy()` doesn't exist? Am I missing something